### ISSUE TITLE
[MNG-6964] Maven version sorting is internally inconsistent

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8070]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** opened **[MNG-6964](https://issues.apache.org/jira/browse/MNG-6964?redirect=false)** and commented

There's a bug where version sorting is inconsistent. This manifests like this:

```java
$ java -jar ~/local/apache-maven/lib/maven-artifact-3.6.3.jar 1-0.alpha 1
Display parameters as parsed by Maven (in canonical form) and comparison result:
1. 1-0.alpha == 1-0.alpha
   1-0.alpha == 1
2. 1 == 1

$ java -jar ~/local/apache-maven/lib/maven-artifact-3.6.3.jar 1-0.beta 1
Display parameters as parsed by Maven (in canonical form) and comparison result:
1. 1-0.beta == 1-0.beta
   1-0.beta == 1
2. 1 == 1

$ java -jar ~/local/apache-maven/lib/maven-artifact-3.6.3.jar 1-0.beta 1-0.alpha
Display parameters as parsed by Maven (in canonical form) and comparison result:
1. 1-0.beta == 1-0.beta
   1-0.beta > 1-0.alpha
2. 1-0.alpha == 1-0.alpha
```

Thus there is no correct total order: `1-0.beta > 1.0.alpha` even though both `1-0.beta` and `1-0.alpha` are equal to `1`.

I think this is likely due to a logical bug where any dot-separated segments following a zero or release (`ga` or `final`) is simply truncated in certain circumstances:

```java
$ java -jar ~/local/apache-maven/lib/maven-artifact-3.6.3.jar 1-0.x 1-0
Display parameters as parsed by Maven (in canonical form) and comparison result:
1. 1-0.x == 1-0.x
   1-0.x == 1-0
2. 1-0 == 1
```

but

```java
$ java -jar ~/local/apache-maven/lib/maven-artifact-3.6.3.jar 1-0.x-1 1-0-1
Display parameters as parsed by Maven (in canonical form) and comparison result:
1. 1-0.x-1 == 1-0.x-1
   1-0.x-1 > 1-0-1
2. 1-0-1 == 1-1
```

---

**Affects:** 3.6.3

**Issue Links:**
- [MRESOLVER-334](https://issues.apache.org/jira/browse/MRESOLVER-334) Maven Resolver's GenericVersionScheme diverges from the spec

- [MNG-6967](https://issues.apache.org/jira/browse/MNG-6967) Improve the command line output from maven-artifact

- [MNG-7700](https://issues.apache.org/jira/browse/MNG-7700) Improper canonicalization of versions

- [MNG-7701](https://issues.apache.org/jira/browse/MNG-7701) Incompatible version sorting changes

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

0 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Dennis Lundberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dennisl)** commented

You can read about how the sorting of versions work here:
https://maven.apache.org/pom.html#version-order-specification

It seems that you are mixing dots '.' and hyphen '-' as if they are interchangeable, but they are not. They have an important meaning when parsing a version number. It says on the above page:

_"This gives a sequence of version numbers (numeric tokens) and version qualifiers (non-numeric tokens) with "." or "-" prefixes."_

In your first console example you use the versions 1-0.beta and 1-0.alpha, but in your conclusion you use 1-0.beta and 1.0.alpha. Note that the difference between dots and hyphens. 1-0.alpha is not the same as 1.0.alpha.

**Comment by jira-importer** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

That was just a typo.  You can see that the result is as I've said up above.  Also it should be highly evident that I understand how version sorting is intended to work.  But I'll fix the typo just to be clear.

**Comment by jira-importer** [comment]
**[David M. Lloyd](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dmlloyd)** commented

I don't understand what you're trying to explain to me.  Yes, there are version numbers that have a valid total order.  But I opened this bug because there are also version numbers that do **not** have a valid total order.  The _social_ question of whether one uses semver or not is completely irrelevant to the problem.  The version comparison is _objectively incorrect_ for the cases I've given above.  That's the beginning, middle, and end of the problem.  I didn't file a bug asking for help, I filed a bug because _there is a bug in the code_.  I hope you can appreciate the difference.

**Comment by jira-importer** [comment]
**[Dennis Lundberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dennisl)** commented

I'm sorry for trying to understand the problem.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [ce27f0ec6184cdee40609817a2796c768d0df261](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=ce27f0ec6184cdee40609817a2796c768d0df261) for `maven-3.8.x` branch.
